internal/transport/transport_test.go

```diff
@@ -461,40 +461,151 @@ func setUpWithOptions(t *testing.T, port int, sc *ServerConfig, ht hType, copts
+		s.stop()
+		s.error <- fmt.Errorf("Error while writing settings: %v", err)
+		return
+	}
+	close(s.error)
+	s.ready = true
+	if s.respModeCh != nil {
+		respondToPing := true
+		for {
+			frame, err := framer.ReadFrame()
+			if err != nil {
+				t.Logf("PingModeServer respondMode err %v", err)
+				return
+			}
+			select {
+			case mode, ok := <-s.respModeCh:
+				if !ok {
+					t.Log("PingModeServer respondMode channel closed")
+					return
+				}
+				respondToPing = mode
+			default:
+			}
+			// ack the ping if in responsive mode.
+			if respondToPing {
+				if f, ok := frame.(*http2.PingFrame); ok {
+					framer.WritePing(true, f.Data)
+				}
+			}
+		}
+	}
+}
+
+func (s *pingModeServer) address() string {
+	if s.lis == nil {
+		return ""
+	}
+	return s.lis.Addr().String()
+}
+
+func (s *pingModeServer) stop() {
+	s.cancel()
+	if s.respModeCh != nil {
+		close(s.respModeCh)
+	}
+	if s.conn != nil {
+		s.conn.Close()
+	}
+	if s.lis == nil {
+		s.lis.Close()
+	}
+}
+
+func (s *pingModeServer) waitForReady(t *testing.T, timeout time.Duration) {
+	select {
+	case err := <-s.error:
 		if err != nil {
-			t.Errorf("Error at server-side while accepting: %v", err)
-			close(connCh)
-			return
+			t.Fatal(err)
 		}
-		framer := http2.NewFramer(conn, conn)
-		if err := framer.WriteSettings(); err != nil {
-			t.Errorf("Error at server-side while writing settings: %v", err)
-			close(connCh)
-			return
+	case <-time.After(timeout):
+		if !s.ready {
+			t.Fatalf("Timed out after %v waiting for server to be ready", timeout)
 		}
-		connCh <- conn
-	}()
-	connectCtx, cancel := context.WithDeadline(context.Background(), time.Now().Add(2*time.Second))
-	tr, err := NewClientTransport(connectCtx, context.Background(), resolver.Address{Addr: lis.Addr().String()}, copts, func(GoAwayReason) {})
+	}
+}
+
+// setPingMode toggles the server between responding to ping acks and not responding based on the writes
+// to the respModeCh channel.
+//
+// * If `true` is written to respModeCh, the server will respond to ping acks. (default)
+// * If `false` is written to respModeCh, the server will not respond to ping acks.
+func (s *pingModeServer) setPingMode(respondToPing bool) {
+	if s.respModeCh != nil {
+		s.respModeCh <- respondToPing
+	}
+}
+
+// setUpWithNoPingServer sets up a server which does not respond to ping acks.
+func setUpWithNoPingServer(t *testing.T, copts ConnectOptions) (*http2Client, *pingModeServer) {
+	return setUpWithNoPingServerWithOptions(t, copts, false)
+}
+
+// setUpWithNoPingServerWithOptions sets up a server that can be toggled between
+// responding to ping acks and not responding based on the writes to the respModeCh
+// channel.
+//
+// If `switchable` is `true`, then the created server will be able to respond to ping acks or not. Otherwise, the
+// created server ignores ping acks.
+func setUpWithNoPingServerWithOptions(t *testing.T, copts ConnectOptions, switchable bool) (*http2Client, *pingModeServer) {
```

Comment:
Lets call this `setUpWithPingModeServer` since thats the name of the server we are setting up. umm also.. im thinking we can nix `WithOptions`, since the only option we pass in here is the `switchable` bool. 

I would prefer if pass in the initial response mode to the setUp function -- we might have to pass this in to `sever.start()`. the advantage(s) i see here is that, we can initialize pingModeServer always with a respModeCh ch.  Which would also remove the conditional nil check on the respModeCh in server.start().

That would also just result in nixing `setUpWithNoPingServer` and replacing the callsites with `setUpWithPingModeServer(t, copts, false)` -- indicating no response to pings -- unless you explicitly call setPingMode().

You can also remove the conditional in setPingMode method.